File: backend/app/schemas/analytics.py
from decimal import Decimal

from pydantic import BaseModel


class NextRenewalInfo(BaseModel):
    service_name: str
    next_billing_date: str
    cost: Decimal


class MostExpensiveInfo(BaseModel):
    service_name: str
    monthly_cost: Decimal


class DashboardOverview(BaseModel):
    total_active_subscriptions: int
    total_monthly_cost: Decimal
    total_yearly_cost: Decimal
    currency: str = "KRW"
    next_renewal: NextRenewalInfo | None = None
    most_expensive: MostExpensiveInfo | None = None


class CategoryBreakdownItem(BaseModel):
    category: str
    total: Decimal
    count: int
    percentage: float


class CategoryBreakdown(BaseModel):
    year: int
    month: int
    breakdown: list[CategoryBreakdownItem]
    total: Decimal


class MonthlyTotal(BaseModel):
    year: int
    month: int
    total: Decimal


class SpendingTrend(BaseModel):
    data: list[MonthlyTotal]

File: frontend/src/types/analytics.ts
export interface DashboardOverview {
  total_active_subscriptions: number;
  total_monthly_cost: number;
  total_yearly_cost: number;
  currency: string;
  next_renewal: {
    service_name: string;
    next_billing_date: string;
    cost: number;
  } | null;
  most_expensive: {
    service_name: string;
    monthly_cost: number;
  } | null;
}

export interface CategoryBreakdownItem {
  category: string;
  total: number;
  count: number;
  percentage: number;
}

export interface CategoryBreakdown {
  year: number;
  month: number;
  breakdown: CategoryBreakdownItem[];
  total: number;
}

export interface MonthlyTotal {
  year: number;
  month: number;
  total: number;
}

export interface SpendingTrend {
  data: MonthlyTotal[];
}
